File: frontend/src/api/schemas.ts
// Hand-written types matching the backend's OpenAPI schema.

export const PAPER_SIZES = ["A3", "A4", "A5", "A6", "Letter", "Legal", "Tabloid"] as const;
export type PaperSize = (typeof PAPER_SIZES)[number];

export interface UploadResponse {
  token: string;
}

export interface StatusResponse {
  token: string;
  status: "pending" | "processing" | "ready" | "error";
  message?: string | null;
}

export interface ContourResponse {
  token: string;
  /** Simplified profile as [[x_mm, y_mm], ...] */
  points: [number, number][];
  /** Pixels per mm from calibration sheet */
  scale: number;
  image_width: number;
  image_height: number;
  spline_points?: [number, number][] | null;
  paper_size: PaperSize;
}

export interface ModelRequest {
  token: string;
  points: [number, number][];
  height?: number;
  format?: "stl" | "step";
  revolve?: boolean;
}

export interface ModelResponse {
  url: string;
  format: string;
  filename: string;
}

// ── History ──────────────────────────────────────────────────────────────────

export interface HistorySaveRequest {
  name?: string;
  points: [number, number][];
  height: number;
  revolve: boolean;
  format: "stl" | "step";
  paper_size: PaperSize;
}

export interface HistoryEntry {
  id: number;
  name: string;
  points: [number, number][];
  height: number;
  revolve: boolean;
  format: string;
  paper_size: string;
  created_at: string;
}

File: backend/app/models/schemas.py
from typing import Literal
from pydantic import BaseModel, Field


class UploadResponse(BaseModel):
    token: str = Field(..., description="Unique processing token for this upload")


class StatusResponse(BaseModel):
    token: str
    status: Literal["pending", "processing", "ready", "error"]
    message: str | None = None


class ContourResponse(BaseModel):
    token: str
    points: list[list[float]] = Field(
        ..., description="Simplified contour as [[x, y], ...] in mm"
    )
    scale: float = Field(..., description="Pixels per mm derived from calibration sheet")
    image_width: int
    image_height: int
    spline_points: list[list[float]] | None = Field(
        None, description="Optional B-spline fitted points"
    )
    paper_size: str = Field(default="A4", description="Calibration sheet size used")


class ModelRequest(BaseModel):
    token: str = Field(..., description="Token from upload step")
    points: list[list[float]] = Field(
        ..., description="Final profile points in mm [[x, y], ...]"
    )
    height: float = Field(default=10.0, gt=0, description="Extrusion height in mm")
    format: Literal["stl", "step"] = Field(default="stl")
    revolve: bool = Field(
        default=False, description="Revolve profile around Y-axis instead of extruding"
    )


class ModelResponse(BaseModel):
    url: str = Field(..., description="Download URL for the generated 3D file")
    format: str
    filename: str


# ── History ──────────────────────────────────────────────────────────────────

class HistorySaveRequest(BaseModel):
    name: str = Field(default="Untitled", max_length=120)
    points: list[list[float]] = Field(..., description="Profile points in mm")
    height: float = Field(default=10.0, gt=0)
    revolve: bool = Field(default=False)
    format: Literal["stl", "step"] = Field(default="stl")
    paper_size: str = Field(default="A4")


class HistoryEntry(BaseModel):
    id: int
    name: str
    points: list[list[float]]
    height: float
    revolve: bool
    format: str
    paper_size: str
    created_at: str
